File: server/buyback.js
'use strict';

const CONDITIONS = new Set([
  'new_sealed',
  'like_new',
  'very_good',
  'used_good',
  'acceptable',
  'defective',
]);

const MAX_AGE_MS = 24 * 60 * 60 * 1000;
const MAX_PRICE_EUR = 10000;

function text(value, max = 240) {
  return String(value || '').trim().slice(0, max);
}

function normalizeBuybackOffer(raw, { now = Date.now() } = {}) {
  if (!raw || typeof raw !== 'object') return null;
  const condition = text(raw.condition, 32);
  const price = Number(raw.price);
  const confidence = Number(raw.match_confidence);
  const checkedAt = Date.parse(raw.checked_at);
  const offerUrl = text(raw.offer_url, 1000);
  let parsedUrl;
  try { parsedUrl = new URL(offerUrl); } catch (_) { return null; }

  if (!CONDITIONS.has(condition)) return null;
  if (!Number.isFinite(price) || price <= 0 || price > MAX_PRICE_EUR) return null;
  if (text(raw.currency, 8) !== 'EUR') return null;
  if (text(raw.price_kind, 40) !== 'indicative_buyback') return null;
  if (!Number.isFinite(confidence) || confidence < 0.9 || confidence > 1) return null;
  if (!Number.isFinite(checkedAt) || checkedAt > now + 5 * 60 * 1000 || now - checkedAt > MAX_AGE_MS) return null;
  if (parsedUrl.protocol !== 'https:') return null;
  if (raw.condition_uncertain === true) return null;

  const providerId = text(raw.provider_id, 80);
  const providerName = text(raw.provider_name, 120);
  const productId = text(raw.product_id, 160);
  const matchedTitle = text(raw.matched_title, 240);
  if (!providerId || !providerName || !productId || !matchedTitle) return null;

  return {
    provider_id: providerId,
    provider_name: providerName,
    product_id: productId,
    matched_title: matchedTitle,
    condition,
    price,
    currency: 'EUR',
    offer_url: parsedUrl.toString(),
    checked_at: new Date(checkedAt).toISOString(),
    price_kind: 'indicative_buyback',
    requires_inspection: raw.requires_inspection !== false,
    match_confidence: confidence,
  };
}

function normalizeBuybackPayload(payload, options) {
  const rawItems = Array.isArray(payload) ? payload : Array.isArray(payload?.items) ? payload.items : [];
  const unique = new Map();
  for (const raw of rawItems) {
    const item = normalizeBuybackOffer(raw, options);
    if (!item) continue;
    const key = `${item.provider_id}\u0000${item.product_id}\u0000${item.condition}`;
    const current = unique.get(key);
    if (!current || item.price > current.price || (item.price === current.price && item.checked_at > current.checked_at)) {
      unique.set(key, item);
    }
  }
  return [...unique.values()].sort((a, b) => b.price - a.price || b.checked_at.localeCompare(a.checked_at));
}

function bestBuybackOffer(items, condition) {
  return items
    .filter((item) => item.condition === condition)
    .reduce((best, item) => (!best || item.price > best.price ? item : best), null);
}

module.exports = { CONDITIONS, MAX_AGE_MS, MAX_PRICE_EUR, normalizeBuybackOffer, normalizeBuybackPayload, bestBuybackOffer };

File: lib/buyback.dart
enum BuybackCondition {
  newSealed,
  likeNew,
  veryGood,
  usedGood,
  acceptable,
  defective,
}

extension BuybackConditionWire on BuybackCondition {
  String get wireValue => switch (this) {
        BuybackCondition.newSealed => 'new_sealed',
        BuybackCondition.likeNew => 'like_new',
        BuybackCondition.veryGood => 'very_good',
        BuybackCondition.usedGood => 'used_good',
        BuybackCondition.acceptable => 'acceptable',
        BuybackCondition.defective => 'defective',
      };

  static BuybackCondition? tryParse(String value) {
    for (final condition in BuybackCondition.values) {
      if (condition.wireValue == value) return condition;
    }
    return null;
  }
}

bool _hasPublicOfferHost(String host) {
  final normalized = host.toLowerCase();
  if (normalized.isEmpty ||
      normalized == 'localhost' ||
      normalized.endsWith('.local') ||
      normalized == '::1') {
    return false;
  }

  if (normalized.contains(':')) {
    // Only globally routable IPv6 destinations belong in user-facing offer links.
    // Block unspecified, loopback, unique-local and link-local ranges.
    return normalized != '::' &&
        !normalized.startsWith('fc') &&
        !normalized.startsWith('fd') &&
        !normalized.startsWith('fe8') &&
        !normalized.startsWith('fe9') &&
        !normalized.startsWith('fea') &&
        !normalized.startsWith('feb');
  }

  final octets = normalized.split('.');
  if (octets.length != 4) return true;
  final ipv4 = octets.map(int.tryParse).toList();
  if (ipv4.any((part) => part == null || part < 0 || part > 255)) return true;

  final a = ipv4[0]!;
  final b = ipv4[1]!;
  final c = ipv4[2]!;
  return !(a == 0 ||
      a == 10 ||
      a == 127 ||
      (a == 100 && b >= 64 && b <= 127) ||
      (a == 169 && b == 254) ||
      (a == 172 && b >= 16 && b <= 31) ||
      (a == 192 && b == 0 && c == 0) ||
      (a == 192 && b == 0 && c == 2) ||
      (a == 192 && b == 168) ||
      (a == 198 && (b == 18 || b == 19)) ||
      (a == 198 && b == 51 && c == 100) ||
      (a == 203 && b == 0 && c == 113) ||
      a >= 224);
}

class BuybackOffer {
  const BuybackOffer({
    required this.providerId,
    required this.providerName,
    required this.productId,
    required this.matchedTitle,
    required this.condition,
    required this.price,
    required this.currency,
    required this.offerUrl,
    required this.checkedAt,
    required this.requiresInspection,
    required this.matchConfidence,
    this.conditionUncertain = false,
  });

  static const double maxComparablePriceEur = 10000;

  final String providerId;
  final String providerName;
  final String productId;
  final String matchedTitle;
  final BuybackCondition condition;
  final double price;
  final String currency;
  final Uri offerUrl;
  final DateTime checkedAt;
  final bool requiresInspection;
  final double matchConfidence;
  final bool conditionUncertain;

  bool get isEligibleForComparison {
    final hasPublicOfferHost = _hasPublicOfferHost(offerUrl.host);

    return !conditionUncertain &&
        providerId.trim().isNotEmpty &&
        providerName.trim().isNotEmpty &&
        productId.trim().isNotEmpty &&
        matchedTitle.trim().isNotEmpty &&
        price.isFinite &&
        price > 0 &&
        price <= maxComparablePriceEur &&
        currency == 'EUR' &&
        offerUrl.hasScheme &&
        offerUrl.scheme == 'https' &&
        hasPublicOfferHost &&
        offerUrl.userInfo.isEmpty &&
        matchConfidence.isFinite &&
        matchConfidence >= 0.9 &&
        matchConfidence <= 1;
  }

  bool isFreshAt(
    DateTime now, {
    Duration maxAge = const Duration(hours: 24),
    Duration futureTolerance = const Duration(minutes: 5),
  }) {
    final age = now.toUtc().difference(checkedAt.toUtc());
    return age >= -futureTolerance && age <= maxAge;
  }

  factory BuybackOffer.fromJson(Map<String, dynamic> json) {
    final condition = BuybackConditionWire.tryParse(json['condition'] as String? ?? '');
    if (condition == null) {
      throw const FormatException('Unsupported buyback condition');
    }

    if (json['price_kind'] != 'indicative_buyback') {
      throw const FormatException('Unsupported buyback price kind');
    }

    final price = (json['price'] as num?)?.toDouble();
    final confidence = (json['match_confidence'] as num?)?.toDouble();
    final checkedAt = DateTime.tryParse(json['checked_at'] as String? ?? '');
    final offerUrl = Uri.tryParse(json['offer_url'] as String? ?? '');

    if (price == null || confidence == null || checkedAt == null || offerUrl == null) {
      throw const FormatException('Incomplete buyback offer');
    }

    return BuybackOffer(
      providerId: json['provider_id'] as String? ?? '',
      providerName: json['provider_name'] as String? ?? '',
      productId: json['product_id'] as String? ?? '',
      matchedTitle: json['matched_title'] as String? ?? '',
      condition: condition,
      price: price,
      currency: json['currency'] as String? ?? '',
      offerUrl: offerUrl,
      checkedAt: checkedAt,
      requiresInspection: json['requires_inspection'] as bool? ?? true,
      matchConfidence: confidence,
      conditionUncertain: json['condition_uncertain'] as bool? ?? false,
    );
  }
}

BuybackOffer? bestComparableBuybackOffer(
  Iterable<BuybackOffer> offers, {
  required BuybackCondition condition,
  DateTime? now,
  Duration maxAge = const Duration(hours: 24),
}) {
  BuybackOffer? best;
  for (final offer in offers) {
    if (offer.condition != condition || !offer.isEligibleForComparison) continue;
    if (now != null && !offer.isFreshAt(now, maxAge: maxAge)) continue;
    if (best == null || offer.price > best.price) best = offer;
  }
  return best;
}
